README.md
## Cacher
It does the hard work so you don't have to!

![Travis](https://travis-ci.org/addisonj/node-cacher.svg)

## What is it?
HTTP Caching implemented as express middleware, with pluggable backends for support for a variety of caching servers (memcached, redis, etc)

## Features
- Set expiry times per-app or per-route. It sets proper headers for client caching
- Avoids the thundering heard by smartly refreshing the cache
- Emits hit or miss events so you can track your hitrate
- Simple invalidations
- Overrides for custom cache keys and dev mode support
- Obeys (some) client Cache-Control headers

## What does it look like?
``` JavaScript
var Cacher = require("cacher")
// use the default in memory cache
var cacher = new Cacher()
// or pass in a different cache client (in this cached memcached) for different backend support
CacherMemcached = require('cacher-memcached')
var cacher = new Cacher(new CacherMemcached('host1:12345'))

// as a global middleware
app.use(cacher.cache('seconds', 30))
// or per route
app.get("/long-cache", cacher.cache('days'), ...)
app.get("/short-cache", cacher.cache('minute'), ...)
// will set no-cache headers for routes that we explicitly want to ignore caching on
app.get("/no-cache", cacher.cache(false), ...)

// Backwards compatible with old cache definitions
app.use(cacher.cacheHourly())
app.get("/long-cache", cacher.cacheDays(2), ...)

// invalidation support
cacher.invalidate('/home')

// listen for events to track cache rate and errors
cacher.on("hit", function(key) {
  console.log("woohoo!")
})
cacher.on("miss", function(key) {
  console.log("doh!")
})
cacher.on("error", function(err) {
  console.log(err)
})

// Dev mode, quickly turn off caching when it gets in the way
var env = process.env.NODE_ENV || 'development'
if (env === 'development') {
  cacher.noCaching = true
}

// override cache key generation for finer grain control
cacher.genCacheKey = function(req) {
  return req.path + req.header('user-agent')
}

// override cache TTL based on response
cacher.genCacheTtl = function(res, origTtl) {
  if (res.statusCode >= 400) {
    return 0
  }
  return origTtl
}
```

## Backends
Currently, Cacher comes bundled with an in-memory cache

Backends are distributed as separate modules:
- cacher-memcached (https://github.com/addisonj/cacher-memcached)
- cacher-redis (https://github.com/addisonj/cacher-redis)


## Backend Client Api
If you want to implement your own backend, the API is really simple:

```JavaScript
// pass whatever options are needed for connection/options
// provide defaults so a client can be fully instantiated with no parameters
function MyBackend(...) {
}

// cb is required, cb(err, cacheObject)
MyBackend.prototype.get = function(key, cb) {}

MyBackend.prototype.set = function(key, cacheObject, ttlInSeconds, [cb]) {}

MyBackend.prototype.invalidate = function(key, [cb]) {}
```

## Testing your backend
Run unit tests using your backend by doing the following:
```Shell
cd Cacher
npm link ../myBackend
CACHER_CLIENT=myBackend npm test
```


CHANGELOG.md
# Changelog

## V3.0.0
- Change behavior for `HEAD` requests to no longer be cached, which could be a breaking change if you are doing something weird with `HEAD` requests (hence the major version bump)
- No API changes





test/testCacher.js
require('http').Agent.maxSockets = 50
var Cacher = require('../lib/Cacher')
var MemoryCache = require('../lib/MemoryClient')
var express = require('express')
var supertest = require('supertest')
var async = require('async')
var assert = require('assert')
var Client = null
var clientConfig = null

if (process.env.CACHER_CLIENT) {
  Client = require(process.env.CACHER_CLIENT)
} else {
  Client = MemoryCache
}

describe('Cacher', function() {
  describe('Instantation', function() {
    it('should instantiate with no parameters and have a proper client', function() {
      var c = new Cacher(new Client())
      assert(c.client instanceof Client)
    })

    it('should not instantiate when passed a client that doesnt adhere to the interface', function() {
      try {
        function BadClient() {}
        var bc = new BadClient()
        var c = new Cacher(bc)
      } catch(e) {
        assert(!!e)
      }
    })
  })

  describe('CalcTtl', function() {
    var c = new Cacher(new Client())

    it('should be able to compute proper ttls from cache values', function() {
      var MIN = 60
      var HOUR = MIN * 60
      var DAY = HOUR * 24
      assert.equal(1, c.calcTtl('second'))
      assert.equal(10, c.calcTtl('seconds', 10))
      assert.equal(MIN, c.calcTtl('minute'))
      assert.equal(MIN*2, c.calcTtl('minutes', 2))
      assert.equal(HOUR, c.calcTtl('hour'))
      assert.equal(HOUR*2, c.calcTtl('hours', 2))
      assert.equal(DAY, c.calcTtl('day'))
      assert.equal(DAY*2, c.calcTtl('days', 2))
      assert.equal(0, c.calcTtl(0))
      assert.equal(0, c.calcTtl(false))
      assert.equal(0, c.calcTtl('days', 0))
    })

  })

  describe('Override Caching', function() {
    var cacher = new Cacher(new Client())
    cacher.noCaching = true

    var app = express()
    app.get('/', cacher.cache('day'), function(req, res) {
      res.send('boop')
    })

    it('shouldnt cache when noCaching is set to true', function(done) {
      supertest(app)
        .get('/')
        .expect(200)
        .expect('Cache-Control', 'no-cache')
        .expect('boop')
        .end(function(err, res) {
          assert.ifError(err)

          supertest(app)
            .get('/')
            .expect(200)
            .expect('Cache-Control', 'no-cache')
            .expect('boop')
            .end(function(err, res) {
              assert.ifError(err)
              done()
            })

        })
    })
  })

  describe('Override TTL based on response', function() {
    var cacher = new Cacher(new Client())
    var app = express()
    var errorCodes = [ 400, 401, 403, 404, 500, 504 ]

    app.get('/error', cacher.cache('minute'), function(req, res) {
      var code = Number(req.param('code')) || 400
      res.send('kaboom', code)
    })

    beforeEach(function() {
      // clear '/error?code=*' entries from cache before each test.
      errorCodes.forEach(function(httpErrorStatusCode) {
        cacher.invalidate('/error?code=' + httpErrorStatusCode)
      })
    })

    it('should return unaltered TTL by default, regardless of response statusCode', function(done) {
      assert.strictEqual(cacher.genCacheTtl({ statusCode: 200 }, 60), 60)

      async.each(
        errorCodes, function(httpErrorStatusCode, next) {
          assert.strictEqual(cacher.genCacheTtl({ statusCode: httpErrorStatusCode }, 60), 60)

          supertest(app)
            .get('/error')
            .query({ code: httpErrorStatusCode })
            .expect(cacher.cacheHeader, 'false')
            .expect(httpErrorStatusCode, 'kaboom')
            .expect('Content-Type', /text/)
            .end(function(err, res) {
              assert.ifError(err)

              supertest(app)
                .get('/error')
                .query({ code: httpErrorStatusCode })
                .expect(cacher.cacheHeader, 'true') // 2nd request is cached
                .expect(httpErrorStatusCode, 'kaboom')
                .expect('Content-Type', /text/)
                .end(function(err, res) {
                  assert.ifError(err)
                  next()
                })
            })
        }, done
      )
    })

    it('should use custom TTL on error responses when genCacheTtl is overridden', function(done) {
      // Override TTL for error responses to be 0 (i.e. not cached).
      cacher.genCacheTtl = function(res, origTtl) {
        if (errorCodes.indexOf(res.statusCode) >= 0) {
          return 0
        }
        return origTtl
      }

      assert.strictEqual(cacher.genCacheTtl({ statusCode: 200 }, 60), 60)

      async.each(
        errorCodes, function(httpErrorStatusCode, next) {
          assert.strictEqual(cacher.genCacheTtl({ statusCode: httpErrorStatusCode }, 60), 0)

          supertest(app)
            .get('/error')
            .query({ code: httpErrorStatusCode })
            .expect(cacher.cacheHeader, 'false')
            .expect(httpErrorStatusCode, 'kaboom')
            .expect('Content-Type', /text/)
            .end(function(err, res) {
              assert.ifError(err)

              supertest(app)
                .get('/error')
                .query({ code: httpErrorStatusCode })
                .expect(cacher.cacheHeader, 'false') // 2nd request is NOT cached
                .expect(httpErrorStatusCode, 'kaboom')
                .expect('Content-Type', /text/)
                .end(function(err, res) {
                  assert.ifError(err)
                  next()
                })
            })
        }, done
      )
    })
  })

  describe('Caching', function() {
    var cacher = new Cacher(new Client())
    var app = require("./fixtures/server")(cacher)

    it('should cache when there is a sufficent ttl', function(done) {
      supertest(app)
        .get('/long')
        .expect(cacher.cacheHeader, 'false')
        .expect(200)
        .expect('long')
        .expect('Content-Type', /text/)
        .end(function(err, res) {
          assert.ifError(err)

          supertest(app)
            .get('/long')
            .expect(cacher.cacheHeader, 'true')
            .expect(200)
            .expect('long')
            .expect('Content-Type', /text/)
            .end(function(err, res) {
              assert.ifError(err)
              done()
            })
        })
    })
    it('shouldnt cache when the ttl expires', function(done) {
      function secondReq() {
        supertest(app)
          .get('/short')
          .expect(cacher.cacheHeader, 'false')
          .expect(200)
          .expect('short')
          .expect('Content-Type', /text/)
          .end(function(err, res) {
            assert.ifError(err)
            done()
          })
      }
      supertest(app)
        .get('/short')
        .expect(cacher.cacheHeader, 'false')
        .expect(200)
        .expect('short')
        .expect('Content-Type', /text/)
        .end(function(err, res) {
          assert.ifError(err)
          setTimeout(secondReq, 1500)
        })
    })

    it('shouldnt cache after an invalidation', function(done) {
      cacher.invalidate('/long', function(err) {
        assert.ifError(err)
        supertest(app)
          .get('/long')
          .expect(cacher.cacheHeader, 'false')
          .expect(200)
          .expect('long')
          .expect('Content-Type', /text/)
          .end(function(err, res) {
            assert.ifError(err)
            supertest(app)
              .get('/long')
              .expect(cacher.cacheHeader, 'true')
              .expect(200)
              .expect('long')
              .expect('Content-Type', /text/)
              .end(function(err, res) {
                assert.ifError(err)
                done()
              })
          })
      })
    })

    it('should be able to dynamically disable ignoring of request headers', function(done) {
      cacher.ignoreClientNoCache = false
      supertest(app)
        .get('/long')
        .set('Cache-Control', 'no-cache')
        .expect(cacher.cacheHeader, 'false')
        .expect(200)
        .expect('long')
        .expect('Content-Type', /text/)
        .end(function(err, res) {
          cacher.ignoreClientNoCache = true
          assert.ifError(err)
          done()
        })
    })

    it('should work with other contentTypes', function(done) {
      supertest(app)
        .get('/json')
        .expect(cacher.cacheHeader, 'false')
        .expect(200)
        .expect({boop: true})
        .expect('Content-Type', /json/)
        .end(function(err, res) {
          assert.ifError(err)

          supertest(app)
            .get('/json')
            .expect(cacher.cacheHeader, 'true')
            .expect(200)
            .expect({boop: true})
            .expect('Content-Type', /json/)
            .end(function(err, res) {
              assert.ifError(err)
              done()
            })
        })

    })

    it('should preserve status codes', function(done) {
     supertest(app)
        .get('/201')
        .expect(cacher.cacheHeader, 'false')
        .expect(201)
        .expect('boop')
        .expect('Content-Type', /text/)
        .end(function(err, res) {
          assert.ifError(err)

          supertest(app)
            .get('/201')
            .expect(cacher.cacheHeader, 'true')
            .expect(201)
            .expect('boop')
            .expect('Content-Type', /text/)
            .end(function(err, res) {
              assert.ifError(err)
              done()
            })
        })
    })
    it('should preserve custom headers', function(done) {
     supertest(app)
        .get('/header')
        .expect(cacher.cacheHeader, 'false')
        .expect(200)
        .expect('header')
        .expect('Content-Type', /text/)
        .expect('X-Custom-Header', 'boop')
        .end(function(err, res) {
          assert.ifError(err)

          supertest(app)
            .get('/header')
            .expect(cacher.cacheHeader, 'true')
            .expect(200)
            .expect('header')
            .expect('Content-Type', /text/)
            .expect('X-Custom-Header', 'boop')
            .end(function(err, res) {
              assert.ifError(err)
              done()
            })
        })
    })
    it('shouldnt cache post request', function(done) {
      supertest(app)
        .post('/post')
        .expect(200)
        .expect('OK')
        .expect('Content-Type', /text/)
        .end(function(err, res) {
          assert(!res.header[cacher.cacheHeader])
          assert.ifError(err)

          supertest(app)
            .post('/post')
            .expect(200)
            .expect('OK')
            .expect('Content-Type', /text/)
            .end(function(err, res) {
              assert(!res.header[cacher.cacheHeader])
              assert.ifError(err)
              done()
            })
        })
    })

    it('shouldnt cache an empty page with HEAD requests', function(done) {
      supertest(app)
        .head('/head')
        .expect(200)
        .expect('Content-Type', /text/)
        .end(function(err, res) {
          assert(!res.header[cacher.cacheHeader])
          assert.ifError(err)
          supertest(app)
            .get('/head')
            .expect(cacher.cacheHeader, 'false')
            .expect('HEAD request')
            .expect(200)
            .expect('Content-Type', /text/)
            .end(function(err, res) {
              assert.ifError(err)
              supertest(app)
                .get('/head')
                .expect(cacher.cacheHeader, 'true')
                .expect('HEAD request')
                .expect(200)
                .expect('Content-Type', /text/)
                .end(function(err, res) {
                  done()
                })
            })
        })
    })

    it('should still work when res.write is used', function(done) {
      supertest(app)
        .get('/write')
        .expect(cacher.cacheHeader, 'false')
        .expect(200)
        .expect('beep|boop')
        .expect('Content-Type', /text/)
        .end(function(err, res) {
          assert.ifError(err)

          supertest(app)
            .get('/write')
            .expect(cacher.cacheHeader, 'true')
            .expect(200)
            .expect('beep|boop')
            .expect('Content-Type', /text/)
            .end(function(err, res) {
              assert.ifError(err)
              done()
            })
        })
    })
    it('should work with attachments', function(done) {
      supertest(app)
        .get('/pdf')
        .expect(cacher.cacheHeader, 'false')
        .expect(200)
        .expect('Content-Type', /pdf/)
        .expect('Content-Length', '7945')
        .end(function(err, res) {
          assert.ifError(err)
          var text = res.text

          supertest(app)
            .get('/pdf')
            .expect(cacher.cacheHeader, 'true')
            .expect(200)
            .expect('Content-Type', /pdf/)
            .expect('Content-Length', '7945')
            .end(function(err, res) {
              assert.ifError(err)
              assert.equal(text, res.text)
              done()
            })
        })
    })
    it('should work with files', function(done) {
      supertest(app)
        .get('/image')
        .expect(cacher.cacheHeader, 'false')
        .expect(200)
        .expect('Content-Type', /png/)
        .expect('Content-Length', '1504')
        .end(function(err, res) {
          assert.ifError(err)
          var text = res.text

          supertest(app)
            .get('/image')
            .expect(cacher.cacheHeader, 'true')
            .expect(200)
            .expect('Content-Type', /png/)
            .expect('Content-Length', '1504')
            .end(function(err, res) {
              assert.ifError(err)
              assert.equal(text, res.text)
              done()
            })
        })
    })
    it('should be able to dynamically disable caching', function(done) {
      cacher.noCaching = true
      supertest(app)
        .get('/long')
        .expect('Cache-Control', 'no-cache')
        .expect(200)
        .end(function(err, res) {
          assert.ifError(err)
          cacher.noCaching = false

          supertest(app)
            .get('/long')
            .expect(cacher.cacheHeader, 'true')
            .expect(200)
            .end(function(err, res) {
              assert.ifError(err)
              done()
            })
        })
    })
    it('should ignore request cache control headers by default', function(done) {
      supertest(app)
        .get('/long')
        .expect(200)
        .end(function(err, res) {
          assert.ifError(err)

          supertest(app)
            .get('/long')
            .set('Cache-Control', 'no-cache')
            .expect(cacher.cacheHeader, 'true')
            .expect(200)
            .end(function(err, res) {
              assert.ifError(err)
              done()
            })
        })
    })

    it('should avoid key collisions when same relative url is used on two different mount points', function(done) {

      var collisionTest = function(expectedCache, next) {
        supertest(app)
          .get('/fooMount/nodupe')
          .expect(cacher.cacheHeader, expectedCache)
          .expect(200, 'foo')
          .end(function(err, res) {
            assert.ifError(err)

            supertest(app)
              .get('/barMount/nodupe')
              .expect(cacher.cacheHeader, expectedCache)
              .expect(200, 'bar')
              .end(function(err, res) {
                assert.ifError(err)
                next()
              })
          })
      }

      async.series([
        function(next) { collisionTest('false', next) },
        function(next) { collisionTest('true', next) },
        function() { done() }
      ])
    })

  })
})


test/fixtures/server.js

var express = require("express")
var path = require("path")
module.exports = function(cacher) {
  var app = express()
  app.get('/long', cacher.cache('day'), function(req, res) {
    res.send('long')
  })
  app.get('/short', cacher.cache('second'), function(req, res) {
    res.send('short')
  })
  app.get('/json', cacher.cache('day'), function(req, res) {
    res.json({boop: true})
  })
  app.get('/201', cacher.cache('day'), function(req, res) {
    res.send("boop", 201)
  })
  app.get('/header', cacher.cache('day'), function(req, res) {
    res.set('X-Custom-Header', 'boop')
    res.send('header')
  })
  app.post('/post',  function(req, res) {
    res.send('OK')
  })
  app.get('/write', cacher.cache('day'), function(req, res) {
    res.type('text')
    res.write('beep|')
    res.end('boop')
  })
  app.get('/pdf', cacher.cache('day'), function(req, res) {
    res.download(path.join(__dirname, 'test.pdf'))
  })
  app.get('/image', cacher.cache('day'), function(req, res) {
    res.sendfile(path.join(__dirname, 'test.png'))
  })
  app.get('/head', cacher.cache('day'), function(req, res) {
    res.send('HEAD request')
  })

  var fooRouter = express.Router();
  var barRouter = express.Router();
  app.use('/fooMount', fooRouter);
  app.use('/barMount', barRouter);
  fooRouter.get('/nodupe', cacher.cache('second', 10), function(req, res) {
    res.send('foo');
  });
  barRouter.get('/nodupe', cacher.cache('second', 10), function(req, res) {
    res.send('bar');
  });

  return app
}


if (module === require.main) {
  var Cacher = require("../../lib/Cacher")
  var MemoryCache = require('../../lib/MemoryClient')
  if (process.env.CACHER_CLIENT) {
    Client = require(process.env.CACHER_CLIENT)
  } else {
    Client = MemoryCache
  }
  var port = process.env.PORT || 3000
  var app = module.exports(new Cacher(new Client()))
  app.listen(port, function() {
    console.log("listening on " + port)
  })
}


test/testMemoryClient.js
var assert = require("assert")
var MemoryClient = require("../lib/MemoryClient")
var D_TTL = 1
describe("MemoryClient", function() {
  var cache = new MemoryClient()
  it("should be able to set a cache value by key and retrieve it", function(done) {
    cache.set("key1", {boop: true}, D_TTL, function(err) {
      assert.ifError(err)
      cache.get("key1", function(err, val) {
        assert.ifError(err)
        assert(val, {boop: true})
        done()
      })
    })
  })

  it("should remove values that expire", function(done) {
    cache.set("expiring", {boop: false}, D_TTL, function(err) {
      assert.ifError(err)
      function checkAfter() {
        cache.get("expiring", function(err, val) {
          assert.ifError(err)
          assert(!val)
          done()
        })
      }

      setTimeout(checkAfter, 1500)
    })
  })

  it("should be able to invalidate a value", function(done) {
    // set a long ttl just in case
    cache.set("invalidate", {boop: true}, 1000, function(err) {
      assert.ifError(err)
      cache.invalidate("invalidate", function(err) {
        assert.ifError(err)
        cache.get("invalidate", function(err, val) {
          assert.ifError(err)
          assert(!val)
          done()
        })
      })
    })
  })


})


lib/Cacher.js
var units = {}
units.second = 1
units.minute = units.second * 60
units.hour = units.minute * 60
units.day = units.hour * 24
units.week = units.day * 7
units.month = units.day * 30
units.year = units.day * 365

// add plural units
Object.keys(units).forEach(function (unit) {
  units[unit + "s"] = units[unit]
})

var STALE_CREATED = 1
var STALE_REFRESH = 2
var GEN_TIME = 30

var HEADER_KEY = 'Cache-Control'
var NO_CACHE_KEY = 'no-cache'
var MAX_AGE_KEY = 'max-age'
var MUST_REVALIDATE_KEY = 'must-revalidate'

var EventEmitter = require('events').EventEmitter
var MemoryClient = require('./MemoryClient')
var util = require('util')

function Cacher(client) {
  // check to make sure they pass in a valid client
  if (client && !client.set && !client.get && !client.invalidate) {
    throw new Error("invalid client")
  } else if (!client) {
    client = new MemoryClient()
  }

  this.client = client
}

util.inherits(Cacher, EventEmitter)

Cacher.prototype.invalidate = function(key, cb) {
  this.client.invalidate(key, cb)
}

Cacher.prototype.cacheDays = function(days) {
  return this.cache('days', days)
}

Cacher.prototype.cacheDaily = function() {
  return this.cache('day')
}

Cacher.prototype.cacheHours = function(hours) {
  return this.cache('hours', hours)
}

Cacher.prototype.cacheHourly = function() {
  return this.cache('hour')
}

Cacher.prototype.cacheMinutes = function(minutes) {
  return this.cache('minutes', minutes)
}

Cacher.prototype.cacheOneMinute = function() {
  return this.cache('minute')
}

Cacher.prototype.noCache = function() {
  return this.cache(false)
}

Cacher.prototype.genCacheKey = function(req) {
  return req.originalUrl
}

Cacher.prototype.genCacheTtl = function(res, origTtl) {
  return origTtl
}

Cacher.prototype.noCaching = false
Cacher.prototype.ignoreClientNoCache = true
Cacher.prototype.cacheHeader = 'X-Cacher-Hit'

Cacher.prototype.calcTtl = function(unit, value) {
  if (unit === 0 || value === 0 || unit === false) return 0
  var unitValue = units[unit]
  if (!unitValue) {
    throw new Error("Unknown unit " + unit)
  }

  if (!value) value = 1
  return unitValue * value
}


function checkNoCache(toCheck) {
  if (!toCheck) return false
  return toCheck.indexOf(NO_CACHE_KEY) > -1
}

Cacher.prototype.cache = function(unit, value) {
  var ttl = this.calcTtl(unit, value),
    self = this

  return function(req, res, next) {
    // set noCaching to true in dev mode to get around stale data when you don't want it
    if (ttl === 0 || self.noCaching) {
      res.header(HEADER_KEY, NO_CACHE_KEY)
      return next()
    }

    // only cache on get
    if (req.method !== 'GET') {
      return next()
    }

    // obey cache-control control headers
    if (!self.ignoreClientNoCache && (checkNoCache(req.header("cache-control")) || checkNoCache(req.header("pragma")))) {
      res.header(self.cacheHeader, false)
      self.emit('miss', self.genCacheKey(req))
      return next()
    }

    var key = self.genCacheKey(req)
    var staleKey = key + ".stale"
    var realTtl = ttl + GEN_TIME * 2

    self.client.get(key, function(err, cacheObject) {
      if (err) {
        self.emit("error", err)
        return next()
      }
      // if the stale key expires, we let one request through to refresh the cache
      // this helps us avoid dog piles and herds
      self.client.get(staleKey, function(err, stale) {
        if (err) {
          self.emit("error", err)
          return next()
        }

        setHeaders(res, ttl)

        if (!stale) {
          self.client.set(staleKey, STALE_REFRESH, GEN_TIME)
          cacheObject = null
        }

        if (cacheObject) {
          self.emit("hit", key)
          return self.sendCached(res, cacheObject)
        }

        self.buildEnd(res, key, staleKey, realTtl, ttl)
        self.buildWrite(res)

        res.header(self.cacheHeader, false)
        next()
        self.emit("miss", key)
      })
    })
  }
}

function setHeaders(res, ttl) {
  res.header(HEADER_KEY, MAX_AGE_KEY + "=" + ttl + ", " + MUST_REVALIDATE_KEY)
}

function appendCache(res, data) {
  if (!data) return
  var buf = data
  if (typeof data === "string") {
    buf = new Buffer(data)
  }
  if (res._responseBody) {
    res._responseBody = Buffer.concat([res._responseBody, buf])
  } else {
    res._responseBody = buf
  }
}

Cacher.prototype.buildEnd = function(res, key, staleKey, realTtl, ttl) {
  var origEnd = res.end
  var self = this

  res.end = function (data) {
    appendCache(res, data)
    var cacheObject = {statusCode: res.statusCode, content: res._responseBody ? res._responseBody.toString("base64") : '', headers: res._headers}
    ttl = self.genCacheTtl(res, ttl)
    if (ttl > 0) {
      self.client.set(key, cacheObject, realTtl, function(err) {
        if (err) {
          self.emit("error", err)
        }
        self.client.set(staleKey, STALE_CREATED, ttl, function(err) {
          if (err) {
            self.emit("error", err)
          }
          self.emit("cache", cacheObject)
        })
      })
    }
    return origEnd.apply(res, arguments)
  }
}

Cacher.prototype.buildWrite = function(res) {
  var origWrite = res.write
  res.write = function (data) {
    appendCache(res, data)
    return origWrite.apply(res, arguments)
  }
}

Cacher.prototype.sendCached = function(res, cacheObject) {
  res.statusCode = cacheObject.statusCode
  for (var header in cacheObject.headers) {
    res.header(header, cacheObject.headers[header])
  }

  res.header(this.cacheHeader, true)

  res.end(new Buffer(cacheObject.content, "base64"))
}

module.exports = Cacher
